src/utils/rlp.ts

```diff
@@ -1,93 +1,229 @@
+        data = val.map((v) => {
+          return v?.toString() || '';
+        });
+      } else if (val instanceof Uint8Array) {
+        data = [val];
+      } else {
+        data = [val?.toString() || ''];
+      }
+
+      singleEncodedValues.push({
+        type: dataType,
+        data,
+      });
+    });
 
-    return val;
-}
\ No newline at end of file
+    encodedValues.push(singleEncodedValues);
+  });
+
+  return encodedValues;
+}
+
+export function analyzeNumber(num: number) {
+  // Convert the number to a string and handle potential negative sign
+  const numStr = Math.abs(num).toString();
+
+  // Check for the presence of a decimal point
+  const decimalIndex = numStr.indexOf('.');
+  const hasDecimal = decimalIndex !== -1;
+
+  // Calculate total digits (excluding the decimal point)
+  const totalDigits = hasDecimal ? numStr.length - 1 : numStr.length;
+
+  // Analysis object to hold the results
+  const analysis = {
+    hasDecimal: hasDecimal,
+    totalDigits: totalDigits,
+    decimalPosition: hasDecimal ? decimalIndex : -1,
+  };
+
+  return analysis;
+}
+
+export function analyzeVariable(val: ValueType): {
+  metadata: [number, number] | undefined;
+  varType: VarType;
+} {
+  if (Array.isArray(val)) {
+    // In Kwil, if there is an array of values, each value in the array must be of the same type.
+    return analyzeVariable(val[0]);
+  }
+
+  let metadata: [number, number] | undefined;
+  // Default to text string
+  // Only other types are null or blob. For client-side tooling, everything else can be sent as a string, and Kwil will handle the conversion.
+  let varType: VarType = VarType.TEXT;
+
+  switch (typeof val) {
+    case 'string':
+      break;
+    case 'number':
+      const numAnalysis = analyzeNumber(val);
+      if (numAnalysis.hasDecimal) {
+        metadata = [numAnalysis.totalDigits, numAnalysis.decimalPosition];
+      }
+      break;
+    case 'boolean':
+      break;
+    case 'object':
+      if (val instanceof Uint8Array) {
+        varType = VarType.BLOB;
+        break;
+      }
+      if (val === null) {
+        varType = VarType.NULL;
+        break;
+      }
+    case 'undefined':
+      varType = VarType.NULL;
+      break;
+    default:
+      throw new Error(
+        `Unsupported type: ${typeof val}. If using a uuid, blob, or uint256, please convert to a JavaScript string.`
+      );
+  }
+
+  return {
+    metadata,
+    varType,
+  };
+}
+
+/**
+ *
+ * @param {ActionBody} actionBody - The body of the action to send. This should use the `ActionBody` interface.
+ * @returns - an array of values to be executed
+ */
+export function encodeArguments(actionBody: ActionBody): EncodedValue[] {
+  // Extract inputs if available
+  const inputs = actionBody?.inputs ? Object.values(actionBody.inputs[0]) : [];
+
+  // Construct encoded values from inputs => (see src/utils/rlp.ts/constructEncodedValues above)
+  const encodedArguments = inputs.length > 0 ? constructEncodedValues([inputs])[0] : [];
```

Comment:
`constructEncodedValues([inputs])[0]` still feels fugly.

What about an approach like below. `encodeNestedArguments` is called when you need to return nested values (i.e. in the ActionBuilder). `encodeSingleArguments` can be called when you need the single `EncodedValue[]` (i.e. in auth.ts).

```typescript
function encodeNestedArguments(preparedAction: ValueType[][]): EncodedValue[][] {
  return preparedAction.map((action) => {
    return encodeSingleArguments(action);
  });
}

function encodeSingleArguments(preparedAction: ValueType[]): EncodedValue[] {
  return preparedAction.map((val) => {
    const { metadata, varType } = analyzeVariable(val);

    const metadataSpread = metadata ? { metadata } : {};

    const dataType: DataType = {
      name: varType,
      is_array: Array.isArray(val),
      ...metadataSpread,
    };

    let data: string[] | Uint8Array[] = [];

    if (Array.isArray(val) && !(val instanceof Uint8Array)) {
      data = val.map((v) => {
        return v?.toString() || '';
      });
    } else if (val instanceof Uint8Array) {
      data = [val];
    } else {
      data = [val?.toString() || ''];
    }

    return {
      type: dataType,
      data,
    };
  });
}
```